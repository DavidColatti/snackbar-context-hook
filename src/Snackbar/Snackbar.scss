.Snackbar {
  &-Content {
    width: 500px;
    display: flex;
    min-height: 45px;
    border-radius: 3px;
    align-items: center;
    padding: 0.4rem 1.5rem;

    &[data-type="success"] {
      background-color: #edf7ed;
    }

    &[data-type="error"] {
      background-color: #fdecea;
    }
  }

  &-Typography {
    color: rgb(87, 87, 87);
    font-size: 0.9rem !important;
  }

  &-Icon {
    margin-right: 1rem;
  }
}
