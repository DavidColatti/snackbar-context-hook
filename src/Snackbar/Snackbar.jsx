import React, { useState, createContext } from "react";
import Snackbar from "@material-ui/core/Snackbar";
import Slide from "@material-ui/core/Slide";
import Paper from "@material-ui/core/Paper";
import Typography from "@material-ui/core/Typography";
import { useSnackBar } from "./useSnackBar";
import "./Snackbar.scss";

const SnackbarContext = createContext({});

const SnackbarProvider = ({ children }) => {
  const snackBarState = useState({
    message: "",
    isOpen: null,
    isSuccessful: false,
  });

  const [
    { isOpen, message, isSuccessful, autoHideDuration },
    setSnackBar,
  ] = snackBarState;

  const handleClose = (e, reason) => {
    if (reason === "clickaway") {
      return;
    }

    setSnackBar({ isOpen: false, isSuccessful: false, message: "" });
  };

  return (
    <SnackbarContext.Provider value={snackBarState}>
      {children}
      <Snackbar
        anchorOrigin={{
          vertical: "bottom",
          horizontal: "center",
        }}
        open={isOpen}
        onClose={handleClose}
        className="Snackbar"
        autoHideDuration={autoHideDuration || 3000}
        TransitionComponent={(props) => <Slide {...props} direction="up" />}
      >
        <Paper
          className="Snackbar-Content"
          data-type={isSuccessful ? "success" : "error"}
        >
          <Typography className="Snackbar-Typography">{message}</Typography>
        </Paper>
      </Snackbar>
    </SnackbarContext.Provider>
  );
};

export { SnackbarProvider, SnackbarContext, useSnackBar };
