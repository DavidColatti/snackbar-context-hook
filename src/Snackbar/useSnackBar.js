import React, { useContext } from "react";
import { SnackbarContext } from "./Snackbar";

export const useSnackBar = (classes) => {
  const { successClassName, errorClassName } = classes || {};
  const [, setSnackbar] = useContext(SnackbarContext);

  const noClasses = classes === undefined;

  const toggleSnackBar = ({ type, message, autoHideDuration }) => {
    const isSuccessful = type === "success";
    const isError = type === "error";
    let iconClassName;

    // if the toggle is successful and we are given a class name
    if (isSuccessful && successClassName) {
      iconClassName = successClassName;

      // if the toggle is successful and there are no specified class names
    } else if (isSuccessful && noClasses) {
      iconClassName = "fas fa-check-circle";

      // if the toggle is an error and we are given a class name
    } else if (isError && errorClassName) {
      iconClassName = errorClassName;

      // if the toggle is an error and there are no specified class names
    } else if (isError && noClasses) {
      iconClassName = "fas fa-exclamation-circle";
    }

    const messageOutput = (
      <>
        <i
          className={`${iconClassName} Snackbar__Icon`}
          data-testid="Snackbar_Icon-Test"
        />
        {message}
      </>
    );

    setSnackbar({
      isOpen: true,
      isSuccessful,
      autoHideDuration,
      message: messageOutput,
    });
  };

  return toggleSnackBar;
};
