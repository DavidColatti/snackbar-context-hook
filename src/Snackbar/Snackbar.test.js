import React from "react";
import { useSnackBar, SnackbarProvider } from "./Snackbar";
import { render, fireEvent, waitFor } from "@testing-library/react";
import "mutationobserver-shim";

// Component to test with
const Component = function ({ type, message, classes }) {
  const snackbar = useSnackBar(classes);

  return (
    <button
      onClick={() =>
        snackbar({
          type,
          message,
          autoHideDuration: 0.1,
        })
      }
    >
      Click
    </button>
  );
};

const classes = {
  successClassName: "successClassName",
  errorClassName: "errorClassName",
};

describe("<SnackbarProvider />", () => {
  it("Should make the context available in the children", () => {
    const { queryByText } = render(
      <SnackbarProvider>
        <Component type="success" message="Success Message" />
      </SnackbarProvider>
    );

    fireEvent.click(queryByText("Click"));

    expect(queryByText("Success Message")).toBeTruthy();
  });

  it("Should have a custom success class name if I pass the parameter to the hook", () => {
    const { queryByText, queryByTestId } = render(
      <SnackbarProvider>
        <Component type="success" message="Success Message" classes={classes} />
      </SnackbarProvider>
    );

    fireEvent.click(queryByText("Click"));

    const icon = queryByTestId("Snackbar_Icon-Test");

    expect(icon.className.includes("successClassName")).toBeTruthy();
  });

  it("Should have a default success icon if I dont pass the parameter to the hook", () => {
    const { queryByText, queryByTestId } = render(
      <SnackbarProvider>
        <Component type="success" message="Success Message" />
      </SnackbarProvider>
    );

    fireEvent.click(queryByText("Click"));

    const icon = queryByTestId("Snackbar_Icon-Test");

    expect(icon.className.includes("fa-check-circle")).toBeTruthy();
  });

  it("Should have a custom error class name if I pass the parameter to the hook", () => {
    const { queryByText, queryByTestId } = render(
      <SnackbarProvider>
        <Component type="error" message="Error Message" classes={classes} />
      </SnackbarProvider>
    );

    fireEvent.click(queryByText("Click"));

    const icon = queryByTestId("Snackbar_Icon-Test");

    expect(icon.className.includes("errorClassName")).toBeTruthy();
  });

  it("Should have a default error icon if I dont pass the parameter to the hook", () => {
    const { queryByText, queryByTestId } = render(
      <SnackbarProvider>
        <Component type="error" message="Error Message" />
      </SnackbarProvider>
    );

    fireEvent.click(queryByText("Click"));

    const icon = queryByTestId("Snackbar_Icon-Test");

    expect(icon.className.includes("fa-exclamation-circle")).toBeTruthy();
  });

  it("Should cover the else statement if not sending any info for the hook function", () => {
    const TestComponent = function () {
      const snackbar = useSnackBar();

      return <button onClick={snackbar}>Click</button>;
    };

    const { queryByText, container } = render(
      <SnackbarProvider>
        <TestComponent />
      </SnackbarProvider>
    );

    fireEvent.click(queryByText("Click"));

    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("useSnackbar", () => {
  it("Should input an icon for a successful snackbar if passed the parameter to the hook", () => {
    const { queryByText } = render(
      <SnackbarProvider>
        <Component
          type="success"
          message="Success Message"
          classes={{
            successClassName: "Snackbar_Success",
            errorClassName: "Snackbar_Error",
          }}
        />
      </SnackbarProvider>
    );

    fireEvent.click(queryByText("Click"));

    expect(queryByText("Success Message")).toBeTruthy();
  });

  it("Should not close the snackbar if clicking outside the component", () => {
    const { queryByText } = render(
      <SnackbarProvider>
        <Component
          type="success"
          message="Success Message"
          classes={{
            successClassName: "Snackbar_Success",
            errorClassName: "Snackbar_Error",
          }}
        />
        <span>Click Outside</span>
      </SnackbarProvider>
    );

    fireEvent.click(queryByText("Click"));
    fireEvent.click(queryByText("Click Outside"));

    expect(queryByText("Success Message")).toBeTruthy();
  });

  it("Should automatically close the snackbar in 3 seconds", async () => {
    const { queryByText } = render(
      <SnackbarProvider>
        <Component
          type="success"
          message="Success Message"
          classes={{
            successClassName: "Snackbar_Success",
            errorClassName: "Snackbar_Error",
          }}
        />
        <span>Click Outside</span>
      </SnackbarProvider>
    );

    fireEvent.click(queryByText("Click"));

    await waitFor(() => {
      expect(queryByText("Success Message")).toBeFalsy();
    });
  });
});
