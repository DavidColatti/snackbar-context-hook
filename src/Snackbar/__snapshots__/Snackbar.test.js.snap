// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SnackbarProvider /> Should cover the else statement if not sending any info for the hook function 1`] = `
<button>
  Click
</button>
`;
